pkg/envoy/ciliumenvoyconfig.go

```diff
@@ -0,0 +1,188 @@
+	}
+	return resources, nil
+}
+
+// UpsertEnvoyResources inserts or updates Envoy resources in 'resources' to the xDS cache,
+// from where they will be delivered to Envoy via xDS streaming gRPC.
+func (s *XDSServer) UpsertEnvoyResources(ctx context.Context, resources Resources) error {
+	log.Debugf("UpsertEnvoyResources: Upserting %d listeners...", len(resources.Listeners))
+	wg := completion.NewWaitGroup(ctx)
+	var revertFuncs xds.AckingResourceMutatorRevertFuncList
+	for i := range resources.Listeners {
+		revertFuncs = append(revertFuncs, s.upsertListener(resources.Listeners[i].Name, resources.Listeners[i], wg, nil))
+	}
+
+	start := time.Now()
+	log.Debug("UpsertEnvoyResources: Waiting for proxy updates to complete...")
+	err := wg.Wait()
+	log.Debug("UpsertEnvoyResources: Wait time for proxy updates: ", time.Since(start))
+
+	// revert all changes in case of failure
+	if err != nil {
+		revertFuncs.Revert(nil)
+		log.Debug("UpsertEnvoyResources: Finished reverting failed xDS transactions")
+	}
+	return err
+}
+
+// UpdateEnvoyResources removes any resources in 'old' that are not
+// present in 'new' and then adds or updates all resources in 'new'.
+// Envoy does not support changing the listening port of an existing
+// listener, so if the port changes we have to delete the old listener
+// and then add the new one with the new port number.
+func (s *XDSServer) UpdateEnvoyResources(ctx context.Context, old, new Resources) error {
+	waitForDelete := false
+	wg := completion.NewWaitGroup(ctx)
+	var revertFuncs xds.AckingResourceMutatorRevertFuncList
+	// Delete old listeners not added in 'new' or if old and new listener have different ports
+	var deleteListeners []*envoy_config_listener.Listener
+	for _, oldListener := range old.Listeners {
+		found := false
+		port := uint32(0)
+		if addr := oldListener.Address.GetSocketAddress(); addr != nil {
+			port = addr.GetPortValue()
+		}
+		for _, newListener := range new.Listeners {
+			if newListener.Name == oldListener.Name {
+				if addr := newListener.Address.GetSocketAddress(); addr != nil && addr.GetPortValue() != port {
+					log.Debugf("UpdateEnvoyResources: %s port changing from %d to %d...", newListener.Name, port, addr.GetPortValue())
+					waitForDelete = true
+				} else {
+					found = true
+				}
+				break
+			}
+		}
+		if !found {
+			deleteListeners = append(deleteListeners, oldListener)
+		}
+	}
+	log.Debugf("UpdateEnvoyResources: Deleting %d, Upserting %d listeners...", len(deleteListeners), len(new.Listeners))
+	for _, listener := range deleteListeners {
+		revertFuncs = append(revertFuncs, s.deleteListener(listener.Name, wg, nil))
+	}
+
+	// Have to wait for deletes to complete before adding new listeners if a listener's port number is changed.
+	if waitForDelete {
+		start := time.Now()
+		log.Debug("UpdateEnvoyResources: Waiting for proxy deletes to complete...")
+		err := wg.Wait()
+		if err != nil {
+			log.Debug("UpdateEnvoyResources: delete failed: ", err)
```

Comment:
should we return `err` if delete fails, or is this error safe to ignore?

Reply:
Envoy likely fails to configure the listener with a new port number if the delete of the old listener here fails. This will then result in an error return from the later `wg.Wait()` in the same function. It could also be that the listener delete failure will not cause the new listener to fail. Either way it will be informative to see if any `UpdateEnvoyResources: delete failed:` lines appear in the logs.

Normally we let Cilium pick up the proxy ports, in which case the port will not change between reconfigurations, and we do not need to wait for the deletes to complete (`waitForDelete` remains `false`).